modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/resource/OpenAPIResourceImpl.java

```diff
@@ -895,6 +896,8 @@ private void _replaceReference(
 		};
 	}
 
+	private static final String[] _URI_SCHEMES = {"http", "https"};
```

Comment:
Here maybe use `Http.HTTP` and `Http.HTTPS` constants?

Reply:
Good point.

Reply:
I was not speaking about LR cloud explicitly, just in general.

When fronting tomcat w/ nginx or httpd, if you terminate SSL there and proxy requests to tomcat, different protocols can be used (https from client to web server, http from web server to tomcat), also the URL hostname and port can be different (www.coke.com would get to the web server, but it will proxy into some internal server(s) names).

So Liferay can generate the right URLs, our code must take advantage of the X-Forwarded headers that can be added to the request handed to tomcat for processing (the web server will populate those with the host, port and protocol used to get to the web tier). By respecting those headers, Liferay generates dynamic URLs which match what the client browser originally issued to get to the site.

Check https://github.com/liferay/liferay-portal/blob/cc85693eba12152385fe9d7354a42d2f81c4ecde/portal-impl/src/com/liferay/portal/util/PortalImpl.java#L2442-L2488 and https://github.com/liferay/liferay-portal/blob/cc85693eba12152385fe9d7354a42d2f81c4ecde/portal-impl/src/com/liferay/portal/util/PortalImpl.java#L6279-L6304 for what I'm talking about.

In fact, if all you do is change the logic to use PortalUtil's getForwardedHost(), getForwardedPort(), and getForwardedProtocol(), or just use PortalUtil's getPortalURL() method, you would have the exact change and correct URL I'm talking about...

Reply:
Got it, we're already [using X-Forward headers](https://github.com/liferay-headless/liferay-portal/pull/740/commits/98b97ee9aec2c94aef3c0cbdc47bb9f6bebef42b#diff-15371b576f26b2c40b7ba851418a02fd8a15011cc03aa936d84c8231516a1fcbR70) when generating openapi URLs, we will apply the same logic to server URL generation. 

Thank you @dnebing!